src/GraphQL/Types/TypeInformation.cs

```diff
@@ -0,0 +1,258 @@
+        /// <see cref="Nullable{T}"/> structs return their underlying type rather than <see cref="Nullable{T}"/>.
+        /// </summary>
+        private static List<(Type Type, NullabilityState Nullable)> Interpret(NullabilityInfo info, bool isInputProperty)
+        {
+            var list = new List<(Type, NullabilityState)>(info.GenericTypeArguments.Length + 1);
+            RecursiveLoop(info);
+            return list;
+
+            void RecursiveLoop(NullabilityInfo info)
+            {
+                if (info.Type.IsGenericType)
+                {
+                    var nullableType = Nullable.GetUnderlyingType(info.Type);
+                    if (nullableType != null)
+                    {
+                        list.Add((nullableType, NullabilityState.Nullable));
+                    }
+                    else
+                    {
+                        list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                    }
+                    foreach (var t in info.GenericTypeArguments)
+                    {
+                        RecursiveLoop(t);
+                    }
+                }
+                else if (info.ElementType != null)
+                {
+                    list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                    RecursiveLoop(info.ElementType);
+                }
+                else
+                {
+                    list.Add((info.Type, isInputProperty ? info.ReadState : info.WriteState));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Applies <see cref="GraphQLAttribute"/> attributes for the specified member to this instance.
+        /// </summary>
+        public virtual void ApplyAttributes()
+        {
+            if (MemberInfo.IsDefined(typeof(System.ComponentModel.DataAnnotations.RequiredAttribute), false))
+            {
+                if (IsList)
+                {
+                    ListIsNullable = false;
+                }
+                else
+                {
+                    IsNullable = false;
+                }
+            }
+
+            var attributes = MemberInfo.GetCustomAttributes(typeof(GraphQLAttribute), false);
+            foreach (var attr in attributes)
+            {
+                ((GraphQLAttribute)attr).Modify(this);
+            }
+        }
+
+        /// <summary>
+        /// Returns a graph type constructed based on the properties set within this instance.
+        /// If <see cref="GraphType"/> is <see langword="null"/>, the graph type is generated via
+        /// <see cref="GraphQL.TypeExtensions.GetGraphTypeFromType(Type, bool, TypeMappingMode)"/>.
+        /// The graph type is then wrapped with <see cref="NonNullGraphType{T}"/> and/or
+        /// <see cref="ListGraphType{T}"/> as appropriate.
+        /// </summary>
+        public virtual Type ConstructGraphType()
```

Comment:
`virtual` here and for `ApplyAttributes`?

Reply:
Yes. I did this to simplify the number of virtual members in AutoRegisteringObjectGraphType. I was going to make this class sealed but after further review it seemed most logical to make them virtual.

Note than in my current code there are a very great number of virtual members. They are all useful, even within my own projects, for different purposes. Now some of the burden is shifted here so if users want to fully customize behavior they might need to derive from this class. 

At the same time that the api is simpler to understand, it is still reasonably easy to replace one piece of functionality without replacing large portions of the codebase.